In [1]:
import pandas as pd

In [2]:
folder = '/Users/BharathiSrinivasan/Documents/Instacart_Data/'

In [3]:
prior = pd.read_csv(folder + "order_products__prior.csv")
train = pd.read_csv(folder + "order_products__train.csv")

In [7]:
prior.shape

(32434489, 4)

In [8]:
train.shape

(1384617, 4)

In [4]:
#We will create a merged dataset from which we can split data into test and train datasets of different sizes

df = pd.concat([prior,train])

In [5]:
orders = pd.read_csv(folder + "orders.csv")
products = pd.read_csv(folder + "products.csv")
dept = pd.read_csv(folder + "departments.csv")

In [6]:
df.dtypes

order_id             int64
product_id           int64
add_to_cart_order    int64
reordered            int64
dtype: object

In [7]:
df1 = pd.merge(df,orders, how = 'left', left_on = 'order_id', right_on = 'order_id')

In [8]:
df2 = pd.merge(df1, products, how = 'left', left_on = 'product_id', right_on = 'product_id')

In [9]:
df_merged = pd.merge(df2, dept, how = 'left', left_on = 'department_id', right_on = 'department_id')

In [10]:
df_merged.head()

order_id  product_id  add_to_cart_order  reordered  user_id eval_set  \
0         2       33120                  1          1   202279    prior   
1         2       28985                  2          1   202279    prior   
2         2        9327                  3          0   202279    prior   
3         2       45918                  4          1   202279    prior   
4         2       30035                  5          0   202279    prior   

   order_number  order_dow  order_hour_of_day  days_since_prior_order  \
0             3          5                  9                     8.0   
1             3          5                  9                     8.0   
2             3          5                  9                     8.0   
3             3          5                  9                     8.0   
4             3          5                  9                     8.0   

            product_name  aisle_id  department_id  department  
0     Organic Egg Whites        86             16  dairy eggs  
1  Michigan Organic Kale        83              4     produce  
2          Garlic Powder       104             13      pantry  
3         Coconut Butter        19             13      pantry  
4      Natural Sweetener        17             13      pantry

In [11]:
df_merged.drop(['eval_set'],axis=1, inplace=True)

In [14]:
#Summary stats
round(df_merged.describe(),2)

order_id   product_id  add_to_cart_order    reordered      user_id  \
count  33819106.00  33819106.00        33819106.00  33819106.00  33819106.00   
mean    1710566.29     25575.51               8.37         0.59    102944.43   
std      987400.76     14097.70               7.14         0.49     59467.33   
min           1.00         1.00               1.00         0.00         1.00   
25%      855413.00     13519.00               3.00         0.00     51435.00   
50%     1710660.00     25256.00               6.00         1.00    102626.00   
75%     2565587.00     37935.00              11.00         1.00    154412.00   
max     3421083.00     49688.00             145.00         1.00    206209.00   

       order_number    order_dow  order_hour_of_day  days_since_prior_order  \
count   33819106.00  33819106.00        33819106.00             31741038.00   
mean          17.14         2.74              13.43                   11.36   
std           17.50         2.09               4.25                    8.94   
min            1.00         0.00               0.00                    0.00   
25%            5.00         1.00              10.00                    5.00   
50%           11.00         3.00              13.00                    8.00   
75%           24.00         5.00              16.00                   15.00   
max          100.00         6.00              23.00                   30.00   

          aisle_id  department_id  
count  33819106.00    33819106.00  
mean         71.22           9.92  
std          38.20           6.28  
min           1.00           1.00  
25%          31.00           4.00  
50%          83.00           9.00  
75%         107.00          16.00  
max         134.00          21.00

* Order ID sequential from 1 to 3,421,083
* Product ID sequential from 1 to 49,688, 134 aisles, 21 departments
* Some exceptinally large orders with upto 145 products in a basket
* 1 to 100 orders of customers present in dataset - a good mix of old and new customers
* Day of week indicated from 0 to 6
* Hour of day as 0 to 23
* Previous order of customer is at a maximum of 30 days and minimum of same day reorder


In [20]:
#Missing values
df_merged[df_merged.isnull().any(axis=1)]

order_id  product_id  add_to_cart_order  reordered  user_id  \
175             20       35430                  1          0   182912   
176             20       47485                  2          0   182912   
177             20        7419                  3          0   182912   
178             20        5112                  4          0   182912   
179             20       33452                  5          0   182912   
180             20       24852                  6          0   182912   
181             20       42633                  7          0   182912   
182             20       27605                  8          0   182912   
320             35       28413                  1          0   202527   
321             35       24532                  2          0   202527   
322             35       14715                  3          0   202527   
323             35       16974                  4          0   202527   
324             35       37980                  5          0   202527   
325             37       35561                  1          0    52564   
326             37       33059                  2          0    52564   
327             37       43721                  3          0    52564   
472             57       21903                  1          0    34016   
473             57       47231                  2          0    34016   
474             57       21137                  3          0    34016   
475             57       43122                  4          0    34016   
476             57       19057                  5          0    34016   
477             57       47209                  6          0    34016   
478             57       48857                  7          0    34016   
617             75       25138                  1          0   106387   
618             75        2361                  2          0   106387   
619             75       10915                  3          0   106387   
620             75       39275                  4          0   106387   
621             75       21616                  5          0   106387   
622             75       12341                  6          0   106387   
623             75       33768                  7          0   106387   
...            ...         ...                ...        ...      ...   
32434191   3421044       39180                  1          0    35140   
32434192   3421044       22035                  2          0    35140   
32434193   3421044       14678                  3          0    35140   
32434194   3421044       34584                  4          0    35140   
32434195   3421044       31506                  5          0    35140   
32434196   3421044       14647                  6          0    35140   
32434197   3421044       39699                  7          0    35140   
32434198   3421044       40584                  8          0    35140   
32434199   3421044       24082                  9          0    35140   
32434200   3421044        9741                 10          0    35140   
32434201   3421044       45866                 11          0    35140   
32434207   3421046       27845                  1          0    48836   
32434208   3421046       33731                  2          0    48836   
32434209   3421046       35547                  3          0    48836   
32434210   3421046       31467                  4          0    48836   
32434211   3421046       45066                  5          0    48836   
32434212   3421046       36412                  6          0    48836   
32434213   3421046       20345                  7          0    48836   
32434214   3421046       37973                  8          0    48836   
32434442   3421077       27966                  1          0   107102   
32434443   3421077       45007                  2          0   107102   
32434444   3421077       26940                  3          0   107102   
32434445   3421077       27845                  4          0   10710

In [21]:
#Missing values in columns other than days since prior order
test = df_merged.drop(['days_since_prior_order'], axis = 1)
test[test.isnull().any(axis=1)]

Empty DataFrame
Columns: [order_id, product_id, add_to_cart_order, reordered, user_id, order_number, order_dow, order_hour_of_day, product_name, aisle_id, department_id, department]
Index: []

In [24]:
# 2 million rows with missing values in days since prior order - Impute with mean
df_merged['days_since_prior_order'] = df_merged['days_since_prior_order'].fillna(value = df_merged['days_since_prior_order'].mean())

In [25]:
#Export dataset
df_merged.to_csv('merged_data.csv')